### PR TITLE
Porting Components from VG: A Hellscape

### PR DESCRIPTION
**Pull request by LadyFowls**
As it says on the tin, essentialy a attempt to port the components system

Also refactors crash_with with stack_trace

### CODE REVIEW

File: code/_helpers/unsorted.dm [hunk 1]
```diff
@@ -1197,3 +1198,19 @@ var/list/wall_items = typecacheof(list(
 		return 0
 	return 1
 
+/proc/CallAsync(datum/source, proctype, list/arguments)
```
Comment (skull132) [inline code/_helpers/unsorted.dm]: InvokeAsync already exists.

File: code/__defines/flags.dm [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+#define NO_DEATHRATTLE_1		16 // Do not notify deadchat about any deaths that occur on this turf.
+#define NO_RUINS_1				32 //Blocks ruins spawning in the area
+//#define CHECK_RICOCHET_1	32		//Same thing as atom flag.
+
+/*
+	These defines are used specifically with the atom/pass_flags bitmask
+	the atom/checkpass() proc uses them (tables will call movable atom checkpass(PASSTABLE) for example)
+*/
+//flags for pass_flags
+#define PASSTABLE		1
```
Comment (Sindorman) [inline code/__defines/flags.dm]: We have these flags already.

File: code/_helpers/unsorted.dm [hunk 3]
```diff
@@ -1006,8 +1006,9 @@ var/list/wall_items = typecacheof(list(
 		color = origin.color
 		set_light(origin.light_range, origin.light_power, origin.light_color)
 
-// call to generate a stack trace and print to runtime logs
-/proc/crash_with(msg)
+
+//gives us the stack trace from CRASH() without ending the current proc.
+/proc/stack_trace(msg)
```
Comment (Sindorman) [inline code/_helpers/unsorted.dm]: Why is it necessary to rename it?

**Comment by Probe1** [comment]
I would recommend the inclusion of the RPED that we ported from TG and made some improvements to.  
https://github.com/vgstation-coders/vgstation13/blob/Bleeding-Edge/code/game/objects/items/weapons/storage/RPED.dm
https://github.com/vgstation-coders/vgstation13/pull/4782

It allows players to screwdriver open machines, use the RPED (rapid parts exchanger device) to swap parts, and screwdriver the machine closed without disassembly allowing our Research/Mechanics players to do mass upgrades without excess effort.  

I'm unfamiliar with Auroras codebase so apologies if this is an unneeded suggestion.

**Comment by Geevies** [comment]
> 
> 
> I would recommend the inclusion of the RPED that we ported from TG and made some improvements to.
> https://github.com/vgstation-coders/vgstation13/blob/Bleeding-Edge/code/game/objects/items/weapons/storage/RPED.dm
> [vgstation-coders/vgstation13#4782](https://github.com/vgstation-coders/vgstation13/pull/4782)
> 
> It allows players to screwdriver open machines, use the RPED (rapid parts exchanger device) to swap parts, and screwdriver the machine closed without disassembly allowing our Research/Mechanics players to do mass upgrades without excess effort.
> 
> I'm unfamiliar with Auroras codebase so apologies if this is an unneeded suggestion.

Far as I know, this already exists in our codebase. It's definitely wonderful! If you've got any more ideas, we'd be glad to hear it.

**Comment by Probe1** [comment]
Oh _god_ I thought this PR was about the minor gameplay mechanic of machine parts.  I'm embarrassed.  I thought it was weird it was bundled with a stack trace refactor but the coffee hadn't kicked in yet.

**Comment by lolman360** [comment]
Might be out of scope, but are you porting/planning to port elements as well?

**Comment by MarinaGryphon** [comment]
`code/modules/mining/mine_turfs.dm:187:error: crash_with: undefined proc`

**Comment by ninjanomnom** [comment]
Where are you porting this from? This is tg components not vg, though it does seem to be out of date in some ways

**Comment by NonQueueingMatt** [comment]
Inactive author, a fuckton of conflicts, reopen when ready.

**Pull request closed without merging**